kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+  private ScheduledExecutorService executor;
+  private ScheduledFuture<?> pollFuture;
+  private Consumer<?, ?> consumer;
+
+  public EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics
+  ) {
+    this(configs, metrics, new Factories() {});
+  }
+
+  EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics,
+      final Factories factories
+  ) {
+    // TODO: make sure to use a read-committed consumer
+    this.configs = consumerConfigs(Objects.requireNonNull(configs));
+    this.metrics = Objects.requireNonNull(metrics);
+    this.factories = Objects.requireNonNull(factories);
+  }
+
+  private Map<String, Object> consumerConfigs(final Map<String, ?> providedConfigs) {
+    final HashMap<String, Object> configs = new HashMap<>(providedConfigs);
+    configs.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    configs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    return Map.copyOf(configs);
+  }
+
+  public ResponsiveConsumer.Listener addForThread(final String threadId) {
+    LOGGER.info("add end offset metrics for thread {}", threadId);
+    final var tm = new ThreadMetrics(threadId);
+    synchronized (this) {
+      assert !threadIdToMetrics.containsKey(threadId);
+      if (threadIdToMetrics.isEmpty()) {
+        initPoller();
+        threadIdToMetrics.put(threadId, tm);
+      }
+    }
+    return tm;
+  }
+
+  public void removeForThread(final String threadId) {
+    final ThreadMetrics tm;
+    synchronized (this) {
+      tm = threadIdToMetrics.remove(threadId);
+      if (tm != null) {
+        tm.cleanup();
+        if (threadIdToMetrics.isEmpty()) {
+          stopPoller();
+        }
+      }
+    }
+  }
+
+  private void pollEndOffsets() {
+    final Set<TopicPartition> partitions = new HashSet<>();
+    final Collection<ThreadMetrics> threadMetrics;
+    synchronized (this) {
+      threadMetrics = List.copyOf(this.threadIdToMetrics.values());
+    }
+    for (final var tm : threadMetrics) {
+      partitions.addAll(tm.endOffsets.keySet());
+    }
+    final var endOffsets = consumer.endOffsets(partitions);
+    threadMetrics.forEach(tm -> tm.update(endOffsets));
+  }
+
+  private void initPoller() {
+    LOGGER.info("init end offsets poller");
+    assert consumer == null;
+    assert executor == null;
+    consumer = factories.createConsumer(configs);
+    executor = factories.createExecutor();
+    pollFuture = executor.scheduleAtFixedRate(this::pollEndOffsets, 0, 30, TimeUnit.SECONDS);
+  }
+
+  private void stopPoller() {
+    LOGGER.info("stopping end offsets poller");
+    try {
+      consumer.close(Duration.ofNanos(0));
+    } catch (final Exception e) {
+      LOGGER.warn("error closing consumer", e);
+    }
+    consumer = null;
+    pollFuture.cancel(true);
+    executor.shutdown();
+    executor = null;
+  }
+
+  private class ThreadMetrics implements ResponsiveConsumer.Listener {
```

Comment:
I think I commented elsewhere, but it would be great if we made this one static and passed in what we need - especially since this is getting passed elsewhere we could leak a reference to the poller

EDIT: ahh I just saw the way this is used with the synchronization, this makes me a bit nervous that we're also leaking the lock on `this` and passing that around. I'm going to write a top-level comment about suggesting a different architecture.

Reply:
done